README.md
---
title: An introduction to knitcitations
author: Carl Boettiger
date: 27 May 2014
bibliography: references.bib
---

[![Build Status](https://travis-ci.org/cboettig/knitcitations.svg?branch=master)](https://travis-ci.org/cboettig/knitcitations)
[![Project Status: Inactive – The project has reached a stable, usable state but is no longer being actively developed; support/maintenance will be provided as time allows.](http://www.repostatus.org/badges/latest/inactive.svg)](http://www.repostatus.org/#inactive)
[![CRAN_Status_Badge](http://www.r-pkg.org/badges/version/knitcitations)](https://cran.r-project.org/package=knitcitations)

<!--
%\VignetteEngine{knitr::rmarkdown}
%\VignetteIndexEntry{An introduction to knitcitations}
-->

<!--
[![Build Status](https://travis-ci.org/cboettig/knitcitations.png)](https://travis-ci.org/cboettig/knitcitations)
-->


knitcitations
=============



- **Author**: [Carl Boettiger](http://www.carlboettiger.info/)
- **License**: [MIT](http://opensource.org/licenses/MIT)
- [Package source code on Github](https://github.com/cboettig/knitcitations)
- [**Submit Bugs and feature requests**](https://github.com/cboettig/knitcitations/issues)


`knitcitations` is an R package designed to add dynamic citations to dynamic documents created with [Yihui's knitr package](https://github.com/yihui/knitr).



Installation 
------------

Install the development version directly from Github 

```r
library(devtools)
install_github("cboettig/knitcitations")
```

Or install the current release from your CRAN mirror with `install.packages("knitcitations")`.  


Quick start: rmarkdown (pandoc) mode
------------------------------------

Start by loading the library.  It is usually good to also clear the bibliographic environment after loading the library, in case any citations are already stored there:  


```r
library("knitcitations")
cleanbib()
```

Set pandoc as the default format:


```r
options("citation_format" = "pandoc")
```

(Note: The old method will eventually be depricated.  For documents using `knitcitations <= 0.5` it will become necessary to set this as `"compatibility"`).  

### Cite by DOI

Cite an article by DOI and the full citation information is gathered automatically. By default this now generates a citation in pandoc-flavored-markdown format. We use the inline command `citep("10.1890/11-0011.1")` to create this citation [@Abrams_2012].  

An in-text citation is generated with `citet`, such as `citet("10.1098/rspb.2013.1372")` creating the citation to @Boettiger_2013.  


### Cite by URL

Not all the literature we may wish to cite includes DOIs, such as [arXiv](http://arxiv.org) preprints, Wikipedia pages, or other academic blogs.  Even when a DOI is present it is not always trivial to locate.  With version 0.4-0, `knitcitations` can produce citations given any URL using the [Greycite API](http://greycite.knowledgeblog.org). For instance, we can use the call `citep("http://knowledgeblog.org/greycite")` to generate the citation to the Greycite tool [@greycite2739].  

### Cite bibtex and bibentry objects directly 

We can also use `bibentry` objects such as R provides for citing packages (using R's `citation()` function): `citep(citation("knitr")` produces [@Xie_2014; @Xie_2013; @Xie_2014a].  Note that this package includes citations to three objects, and pandoc correctly avoids duplicating the author names.  In pandoc mode, we can still use traditional pandoc-markdown citations like `@Boettiger2013` which will render as @Boettiger2013 without any R code, provided the citation is already in the `.bib` file we name (see below).


### Re-using Keys

When the citation is called, a key in the format `FirstAuthorsLastName_Year` is automatically created for this citation, so we can now continue to cite this article without remembering the DOI, using the command `citep("Abrams_2012")` creates the citation [@Abrams_2012] without mistaking it for a new article.  






### Displaying the final bibliography

At the end of the document, include a chunk containing the command:


```r
write.bibtex(file="references.bib")
```

Use the chunk options `echo=FALSE` and `message=FALSE` to hide the chunk command and output.  

This creates a Bibtex file with the name given.  [Pandoc](http://johnmacfarlane.net/pandoc) can then be used to compile the markdown into HTML, MS Word, LaTeX, PDF, or many other formats, each with the desired journal styling. Pandoc is now integrated with [RStudio](http://rstudio.com) through the [rmarkdown](http://rmarkdown.rstudio.com) package.  Pandoc appends these references to the end of the markdown document automatically.  In this example, we have added a yaml header to our Rmd file which indicates the name of the bib file being used, and the optional link to a [CSL](https://github.com/citation-style-language/styles) stylesheet which formats the output for the ESA journals:

```yaml
---
bibliography: "references.bib"
csl: "ecology.csl"
output:
  html_document
---
```


Then calling `rmarkdown::render("tutorial.Rmd")` from R on the tutorial compiles the output markdown, with references in the format of the ESA journals.  

# References















NEWS
NEWS 
====

For more fine-grained list of changes or to report a bug, consult 

* [The issues log](https://github.com/cboettig/knitcitations/issues)
* [The commit log](https://github.com/cboettig/knitcitations/commits/master)

Versioning
----------

Releases will be numbered with the following semantic versioning format:

<major>.<minor>.<patch>

And constructed with the following guidelines:

* Breaking backward compatibility bumps the major (and resets the minor 
  and patch)
* New additions without breaking backward compatibility bumps the minor 
  (and resets the patch)
* Bug fixes and misc changes bumps the patch
* Following the RStudio convention, a .99 is appended after the patch
  number to indicate the development version on Github.  Any version
  Coming from Github will now use the .99 extension, which will never
  appear in a version number for the package on CRAN. 


For more information on SemVer, please visit http://semver.org/.

v1.0.8
------

2017-06-24

- Avoid example that depends on XML (no longer inherited dependency from RefManageR)
- add badges to README


v1.0.7
------

- Avoid failures in CRAN tests do to network connectivity issues.
- Update importFrom directive for previously-considered-base package, `utils`

v1.0.5
------

- Add template file for the MIT license, as requested from CRAN. We include the entire MIT License


v1.0.4
------

- Add error handling when Greycite API cannot be reached

v1.0.3
--------

- Update vignette to reflect changes to `rmarkdown` and pandoc (see [#67](https://github.com/cboettig/knitcitations/issues/67))
- Fixed missing citation in vignette building (knitcitations default format for bibtex keys now has an underscore, `LastName_Year`)

v1.0.2
---------

- Fix passing of arguments to `cite_options` that was causing many options to be ignored. (See [#63](https://github.com/cboettig/knitcitations/issues/63))
- Added \donttest on examples needing internet connection (resource may not be available)
- Licence migrated to MIT

v1.0-1
------

This version is a ground-up rewrite of knitcitations, providing a more powerful interface while also streamlining the back end, mostly by relying more on external libraries for knitty gritty. While an effort has been made to preserve the most common uses, some lesser-used functions or function arguments have been significantly altered or removed. Bug reports greatly appreciated.  

- `citet`/`citep` now accept more options.  In addition to the four previously supported options (DOI, URL, bibentry or bibkey (of a previously cited work)), these now accept a plain text query (used in a CrossRef Search), or a path to a PDF file (which attempts metadata extraction).  

- Citation key generation is now handled internally, and cannot be configured just by providing a named argument to `citet`/`citep`.  

- The `cite` function is replaced by `bib_metadata`.  This function takes any argument to `citet`/`citep` as before (including the new arguments), see docs.  

- Linked inline citations now use the configuration: cite_options(style="markdown", hyperlink="to.doc") provides a link to the DOI or URL of the document, using markdown format.  

- Support for cito and tooltip have been removed.  These may be restored at a later date.  (The earlier implementation did not appropriately abstract the use of these features from the style/formatting of printing the citation, making generalization hard.  

- `bibliography` now includes CSL support directly for entries with a DOI using the `style=` argument.  No need to provide a CSL file itself, just the name of the journal (or rather, the name of the corresponding csl file: full journal name, all lower case, spaces as dashes). See https://github.com/cboettig/knitcitations/issues/38 

- `bibliography` formatting has otherwise been completely rewritten, and no longer uses print_markdown, print_html, and print_rdfa methods.  rdfa is no longer available, and other formats are controlled through cite_options.  For formal publication pandoc mode is recommended instead of `bibliography`.  



v0.6-2
------

- Apply accents to names for in-text citations (issue [#51](https://github.com/cboettig/knitcitations/issues/51))
- Use first name if author last name is not available, e.g. `citet(citation())` (issue [#55](https://github.com/cboettig/knitcitations/issues/55))


v0.6-1
------

- Bugfix in default `citep` method (caused by missing `page` argument)

v0.6-0
------

- Pandoc formatting is introduced as the prefered way to handle citations.  The `.Rmd` files remain as before, importing citation data on the fly with `citep()` and `citet` taking DOIs, URLs, or bibentries as arguments, but render in pandoc's markdown. The citations used are written to a bibtex file and the user must use pandoc (either directly or through RStudio's [rmarkdown](http://rmarkdown.rstudio.com)) to format the citations appropriately.  See the updated vignette & README. This method has to be enabled with `options(citation_format = "pandoc")` at present, otherwise compatibility mode is enabled by default.  (see [#57](https://github.com/cboettig/knitcitations/issues/57)) 


v0.5-0
------
* html print method now provides support for bulleted lists, see [#41](https://github.com/cboettig/knitcitations/issues/41)


v0.4-7
------
* Hard-wrap roxygen documentation to avoid line overflows in pdf manual

v0.4-6
------
* html print method now handled directly in knitcitations just like the
  markdown print method, with ordering, etc, since default html printing 
  method sucks
* ~~Tooltip provides HTML formatting, with link (issue [#37](https://github.com/cboettig/knitcitations/issues/37))~~
* HTML formatting inside tooltip (data-html="true") removed, since incompatible
  with Pandoc parser. [pandoc/#831](https://github.com/jgm/pandoc/issues/831) 

v0.4-5
------

* Provide page range as option to `citep` and `citet` (see issue [#32](https://github.com/cboettig/knitcitations/issues/32))

R/bibtex_io.R
#' read in bibtex and use key as list names
#'
#' @param bibfile a bibtex .bib file
#' @return a list of citation information
#' @details this differs from read.bib in that the list is named.
#' this allows one to use citep(bib[c("key1", "key2")]
#' @import bibtex
#' @export
#' @seealso read.bib citep citet
read.bibtex <- function(bibfile){
  bibs <- read.bib(bibfile)
  keys <- lapply(bibs, function(entry) entry$key)
  names(bibs) <- keys
  bibs
}

#' write a bibtex file  
#'
#' @param entry a 'bibentry' object, list of bibentry objects, or a character vector of package names
#' @param file output bibtex file. Will automatically append '.bib' if not added.
#'   if 'NULL' will use stdout.  
#' @param append a logical indicating that bibtex entries be added the the file.  
#'  If FALSE (default), the file is overwritten.  
#' @param verbose a logical to toggle verbosity. If 'file=NULL', verbosity is
#'  forced off.
#' @return a list of citation information, invisibly
#' @details This function is simply a wrapper to the function write.bib 
#' by Renaud Gaujoux.  Though that function has been added to the 'bibtex'
#' package by Romain Francois (a more sensible place to find it), that version
#' was not avialble on CRAN at the time of writing.  
#' 
#' The 'knitcitations' package automatically extends the use of this function to
#' be able to write bibtex files from a string of DOIs, making it valuable for 
#' purposes beyond the citation of packages.  
#' 
#' @import bibtex pkgmaker
#' @examples
#'  write.bibtex(c('bibtex', 'knitr', 'knitcitations'), file="example.bib")
#'  refs <- lapply(c("10.1111/j.1461-0248.2005.00827.x","10.1890/11-0011.1"), ref)
#'  write.bibtex(refs, file="refs.bib")
#' 
#' @export
#' @seealso read.bib citep citet
write.bibtex <- function(entry, file="knitcitations.bib", append=FALSE, verbose=TRUE){
  if(is(entry, "list") & is(entry[[1]], "bibentry")){    
    write.bib(entry[[1]], file=file, append=append, verbose=verbose)
    for(i in 2:length(entry))
      write.bib(entry[[i]], file=file, append=TRUE, verbose=verbose)
  } else {
    write.bib(entry=entry, file=file, append=append, verbose=verbose)
  }
}



